### DISCUSSION TITLE
How annotations/bindings are determined

### DISCUSSION THREAD

**Discussion by Trainmaster9977** [ScanMountGoat/Smush-Material-Research#34] [Q&A] [answered]
Didn’t realize this was a thing, so apologies for using issues earlier.

In short, I want to ask how what binds to say, fp_c9 is determined. The buffer bindings in the json and from what I can tell, the uniform data in the shader itself, don’t seem to match up. So how were you able to match them? I mean you can do it manually if you compare to renderdoc but you obviously aren’t going to do that for every single file. So how were you able to do it?

**Comment by ScanMountGoat** [comment]
Buffers and also textures are referenced using their binding index. There's not an obvious mapping between the nushdb data and the bindings in the shaders. The names in Ryujinx like "fp_c9_data" are just generated automatically based on integer handles, indexes, etc from the compiled code. 

There are two main ways to refer to resources in modern graphics APIs. You either explicitly set the handle/index in the shader code or you query the index in code using something like `getResourceLocation("myResource")` where `getResourceLocation` is some function provided by the graphics API. The variable names are just for convenience. I would strongly recommend learning some OpenGL first, since this isn't something you can figure out just by staring at the shader code. Search for "opengl uniform block" or "opengl uniform buffer object".
https://www.khronos.org/opengl/wiki/Layout_Qualifier_(GLSL)#Binding_points

The relevant code from smush_materials. The materials seem to always be in fp_c9 or vp_c9. It should be possible in theory to map all the inputs to the shaders, but it would require a lot of testing to validate.
https://github.com/ScanMountGoat/Smush-Material-Research/blob/925fd435a9fe0e0f788cffaf5cc71a1c86a47c46/smush_materials/src/main.rs#L756-L773

**Reply by ScanMountGoat** [reply, marked as answer]
The bindings for uniforms, input/output attributes, textures, and buffers have been applied to the recently uploaded annotated shader dump. For anyone curious as to how the bindings work, please consult the annotation code in smush_materials. The shdr_data module in ssbh_data was also updated recently with new information. 

There are some differences between the shaders generated by Ryujinx at runtime and those generated with Ryujinx.ShaderTools that was causing the confusion. I don't have a way to test every single shader, but the new logic is straightforward and matches up with with manually mapping values in Ryujinx and looking at decompiled and dissassembled code.

**Comment by Trainmaster9977** [comment]
In other words, you haven't been able to figure out _how/why_ it usually maps to those numbers specifically (even using envydis to convert the binary to assembly, the code uses the same numbers) , but you do know that the same things tend to be in the same spots in practice?

**Reply by ScanMountGoat** [reply]
Correct. The shaders are precompiled, so it needs to be deterministic. Figuring it out isn't relevant to the applications I develop, so I haven't looked into it much. It's only really required if you want to use the decompiled shaders in an application without writing your own from scratch. Fully annotating the shaders would be cool at some point, but it's not something I plan on implementing. Hopefully this helps someone else get started if they choose to look into it.

**Reply by Trainmaster9977** [reply]
I plan to edit the shaders and put them back into the game- can't be done if I decompile to GLSL but if I do so to assembly it's pretty easy going. While I can just use renderdoc to check things manually, that's a lot of work if I want to do this for multiple. ESP since I plan to write a tutorial on editing them once I get my first few out the door. But seems the manual way will have to do for now. Thanks!

(Apologies in advance for the requests you're going to get regarding custom shaders in SSBH :P. Don't feel obligated to support em.)
